fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctRewriteRule.java

```diff
@@ -0,0 +1,195 @@
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalOlapScanOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+import org.jetbrains.annotations.NotNull;
+import org.spark_project.guava.collect.ImmutableMap;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/*
+ *
+ * Rewrite count(distinct xx) group by x when distinct columns satisfy scan distribution
+ *
+ * e.g.
+ * select count(distinct xx) from t group by x
+ * ->
+ * select count(xx) from (select x, xx from t group by x, xx) tt group by x;
+ *
+ */
+public class GroupByCountDistinctRewriteRule extends TransformationRule {
+    // multi-stage distinct function mapping
+    // origin aggregate -> 2nd stage
+    private static final Map<String, String> DISTINCT_FUNCTION_TRANS = ImmutableMap.<String, String>builder()
+            .put(FunctionSet.COUNT, FunctionSet.COUNT)
+            .put(FunctionSet.SUM, FunctionSet.SUM)
+            .put(FunctionSet.MULTI_DISTINCT_SUM, FunctionSet.SUM)
+            .put(FunctionSet.MULTI_DISTINCT_COUNT, FunctionSet.COUNT)
+            .build();
+
+    // multi-stage other function mapping
+    // origin aggregate -> Pair<1st stage, 2nd stage>
+    // these aggregate function can be support multi-stage
+    private static final Map<String, Pair<String, String>> OTHER_FUNCTION_TRANS =
+            ImmutableMap.<String, Pair<String, String>>builder()
+                    .put(FunctionSet.COUNT, Pair.create(FunctionSet.COUNT, FunctionSet.SUM))
+                    .put(FunctionSet.MAX, Pair.create(FunctionSet.MAX, FunctionSet.MAX))
+                    .put(FunctionSet.MIN, Pair.create(FunctionSet.MIN, FunctionSet.MIN))
+                    .put(FunctionSet.SUM, Pair.create(FunctionSet.SUM, FunctionSet.SUM))
+                    .put(FunctionSet.HLL_UNION, Pair.create(FunctionSet.HLL_UNION, FunctionSet.HLL_UNION))
+                    .put(FunctionSet.NDV, Pair.create(FunctionSet.HLL_UNION, FunctionSet.NDV))
+                    .put(FunctionSet.BITMAP_UNION, Pair.create(FunctionSet.BITMAP_UNION, FunctionSet.BITMAP_UNION))
+                    .put(FunctionSet.BITMAP_UNION_COUNT,
+                            Pair.create(FunctionSet.BITMAP_UNION, FunctionSet.BITMAP_UNION_COUNT))
+                    .put(FunctionSet.PERCENTILE_UNION,
+                            Pair.create(FunctionSet.PERCENTILE_UNION, FunctionSet.PERCENTILE_UNION))
+                    .build();
+
+    public GroupByCountDistinctRewriteRule() {
```

Comment:
What if there is project above scan operator?

Reply:
project has been merged